manual/case_sensitivity/README.md

```diff
@@ -0,0 +1,129 @@
+does not capture the form; when a method processed an identifier, it always had to know where it
+came from and what form it was in, and possibly convert it. This led a lot of internal complexity,
+and recurring bugs.
+
+To address this issue, driver 4+ uses a wrapper: [CqlIdentifier]. Its API methods are always
+explicit about the form:
+
+```java
+CqlIdentifier caseInsensitiveId = CqlIdentifier.fromCql("FooBar");
+System.out.println(caseInsensitiveId.asInternal());             // foobar
+System.out.println(caseInsensitiveId.asCql(/*pretty=*/ false)); // "foobar"
+System.out.println(caseInsensitiveId.asCql(true));              // foobar
+
+// Double-quotes need to be escaped inside Java strings
+CqlIdentifier caseSensitiveId = CqlIdentifier.fromCql("\"FooBar\"");
+System.out.println(caseSensitiveId.asInternal());               // FooBar
+System.out.println(caseSensitiveId.asCql(true));                // "FooBar"
+System.out.println(caseSensitiveId.asCql(false));               // "FooBar"
+```
+
+*Side note: as shown above, `asCql` has a pretty-printing option that omits the quotes if possible.
+This looks nicer, but is slightly more expensive because it requires parsing the string.*
+
+The driver API uses `CqlIdentifier` whenever it produces or consumes an identifier. For example:
+
+* getting the keyspace from a table's metadata: `CqlIdentifier keyspaceId =
+  tableMetadata.getKeyspace()`;
+* setting the keyspace when building a session: `CqlSession.builder().withKeyspace(keyspaceId)`.
+
+For "consuming" methods, string overloads are also provided for convenience, for example
+`SessionBuilder.withKeyspace(String)`.
+
+* getters and setters of "data container" types [Row], [UdtValue], and [BoundStatement] follow
+  special rules described [here][AccessibleByName] (these methods are treated apart because they are
+  generally invoked very often, and therefore avoid to create `CqlIdentifier` instances internally); 
+* in other cases, the string is always assumed to be in CQL form, and converted on the fly with
+  `CqlIdentifier.fromCql`. 
+
+[CqlIdentifier]:    http://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/CqlIdentifier.html
+[Row]:              http://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/cql/Row.html
+[UdtValue]:         http://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/data/UdtValue.html
+[BoundStatement]:   http://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/cql/BoundStatement.html
+[AccessibleByName]: http://docs.datastax.com/en/drivers/java/4.0/com/datastax/oss/driver/api/core/data/AccessibleByName.html
+
+### Good practices
+
+As should be clear by now, case sensitivity introduces a lot of extra (and arguably unnecessary)
+complexity.
+
+The Java driver's team recommendation is:
+
+> **Always use case-insensitive identifiers in your data model.**
```

Comment:
Amen 😆

Reply:
Change font size to 300?

Reply:
Add a few giant red arrows and make it blink too 🤣